### PR TITLE
Simplify Balancer pool initialization.

### PR DESCRIPTION
**Pull request by nlordell**
This PR simplifies Balancer pool initialization to get rid of some duplicate code. Specifically, we no longer eagerly split `graph_api::PoolData` into one field for each factory. We now get a list of pool data and convert it into each of the factory's `PoolInfo` type (i.e. immutable pool information) when initializing each factory's registry.

Again this is in the spirit of making adding new pool types easier (so the boiler plate of splitting out the `PoolData` for a specific factory is no longer needed).

### Test Plan

Added unit test for new registered pool grouping logic.

### CODE REVIEW

**Review by bh2smith — APPROVED** [review]
Love to see a bunch of redundant stuff gone!

File: crates/shared/src/sources/balancer_v2/graph_api.rs [hunk 1]
```diff
@@ -318,6 +340,45 @@ mod tests {
         );
     }
 
+    #[test]
+    fn splits_pools_by_factory() {
```
[the remark this answers is not in the document]
Comment (bh2smith) [reply crates/shared/src/sources/balancer_v2/graph_api.rs:344]: I see you liked my suggestion here.

[the remark this answers is not in the document]
Comment (nlordell) [reply crates/shared/src/sources/balancer_v2/graph_api.rs]: Done.

**Review by fleupold — APPROVED** [review]
Nice, this refactoring is really starting to take form. So much 🟥  code ✂️...

**Review by bh2smith — APPROVED** [review]

**Pull request merged**